daemon/listener/listener.js

```diff
@@ -0,0 +1,303 @@
+const process = require('process')
+const http = require('http')
+const urllib = require('url')
+const Origin = require('../../dist/index') // Will eventualy be the origin npm package
+const Web3 = require('web3')
+
+const web3Provider = new Web3.providers.HttpProvider('http://localhost:8545')
+const web3 = new Web3(web3Provider)
+const o = new Origin({ web3 })
+
+// Origin Listener
+// ---------------
+// An at-least-once event listener for origin events (May deliver an event more than once).
+// Make sure your webhook endpoint is idempotent!
+// Sends events along with all the related data.
+// Designed to be infura compatible - API calls only, no subscriptions.
+//
+// To use:
+// - Run `npm start run` to setup a local IPFS and blockchain, and run tests
+// - In another terminal run `node scripts/listener.js`
+// - Or `node scripts/listener.js --webhook=http://localhost/originevents/`
+//
+// Todo
+// - Allow configuring web3 endpoint and IPFS endpoint
+// - Handle errors
+// - Persist starting point
+// - Handle blockchain splits/winners
+// - Include current-as-of block numbers in POSTs
+// - Perhaps send related data as it was at the time of the event, not crawl time
+
+// -----------------------------
+// Section 1: Follow rules
+// -----------------------------
+
+// Helper functions
+const generateListingId = log => {
+  return [log.networkId, log.contractVersionKey, log.decoded.listingID].join(
+    '-'
+  )
+}
+const generateOfferId = log => {
+  return [
+    log.networkId,
+    log.contractVersionKey,
+    log.decoded.listingID,
+    log.decoded.offerID
+  ].join('-')
+}
+const getListingDetails = async log => {
+  return {
+    listing: await o.marketplace.getListing(generateListingId(log))
+  }
+}
+const getOfferDetails = async log => {
+  return {
+    listing: await o.marketplace.getListing(generateListingId(log)),
+    offer: await o.marketplace.getOffer(generateOfferId(log))
+  }
+}
+
+// Rules for acting on events
+// Adding a rule here makes the listener listen for the event.
+const LISTEN_RULES = {
+  v00_MarketplaceContract: {
+    ListingCreated: getListingDetails,
+    ListingUpdated: getListingDetails,
+    ListingWithdrawn: getListingDetails,
+    ListingData: getListingDetails,
+    ListingArbitrated: getListingDetails,
+    OfferCreated: getListingDetails,
+    OfferWithdrawn: getOfferDetails,
+    OfferAccepted: getOfferDetails,
+    OfferDisputed: getOfferDetails,
+    OfferRuling: getOfferDetails,
+    OfferFinalized: getOfferDetails,
+    OfferData: getOfferDetails
+  },
+  v01_MarketplaceContract: {
```

Comment:
The code at this point is somewhat teaching code for explaining how the listener works. I think there's value for now in optimizing for making the structure very explicit. Once we get more versions, it will make sense to merge some of them.

Reply:
a ok makes sense